src/Symfony/Component/Security/Core/Authentication/Provider/LdapQueryAuthenticationProvider.php

```diff
@@ -0,0 +1,98 @@
+use Symfony\Component\Ldap\Exception\ConnectionException;
+
+/**
+ * LdapQueryAuthenticationProvider authenticates a user against an LDAP server.
+ *
+ * The only way to check user credentials is to first find the DN via query and then bind
+ *
+ * @author Charles Sarrazin <charles@sarraz.in>
+ * @author Lukas Kahwe Smith <smith@pooteeweet.org>
+ */
+class LdapQueryAuthenticationProvider extends UserAuthenticationProvider
+{
+    private $userProvider;
+    private $ldap;
+    private $dnString;
+    private $queryString;
+
+    /**
+     * Constructor.
+     *
+     * @param UserProviderInterface $userProvider               A UserProvider
+     * @param UserCheckerInterface  $userChecker                A UserChecker
+     * @param string                $providerKey                The provider key
+     * @param LdapInterface         $ldap                       A Ldap client
+     * @param string                $dnString                   A string with the query DN
+     * @param string                $queryString                A string used to create the query query
+     * @param bool                  $hideUserNotFoundExceptions Whether to hide user not found exception or not
+     */
+    public function __construct(UserProviderInterface $userProvider, UserCheckerInterface $userChecker, $providerKey, LdapInterface $ldap, $dnString, $queryString = '{username}', $hideUserNotFoundExceptions = true)
+    {
+        parent::__construct($userChecker, $providerKey, $hideUserNotFoundExceptions);
+
+        $this->userProvider = $userProvider;
+        $this->ldap = $ldap;
+        $this->dnString = $dnString;
+        $this->queryString = $queryString;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function retrieveUser($username, UsernamePasswordToken $token)
+    {
+        if (AuthenticationProviderInterface::USERNAME_NONE_PROVIDED === $username) {
+            throw new UsernameNotFoundException('Username can not be null');
+        }
+
+        return $this->userProvider->loadUserByUsername($username);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function checkAuthentication(UserInterface $user, UsernamePasswordToken $token)
+    {
+        $username = $token->getUsername();
+        $password = $token->getCredentials();
+
+        if ('' === $password) {
+            throw new BadCredentialsException('The presented password must not be empty.');
+        }
+
+        try {
+            $username = $this->ldap->escape($username, '', LdapInterface::ESCAPE_DN);
+            $query = str_replace('{username}', $username, $this->queryString);
+
+            $query = $this->ldap->query($this->dnString, $query);
+            $result = $query->execute();
+            if (1 !== $result->count()) {
+                throw new BadCredentialsException('The presented password is invalid.');
+            }
+
+            $this->ldap->bind($result->getIterator()->current()->getDn(), $password);
```

Comment:
this could probably be rewritten as `$this->ldap->bind($result[0]->getDn(), $password);`